### PR TITLE
IA-2494 Org Unit Change Request feedback

### PR DESCRIPTION
**Pull request by kemar**
Org Unit Change Request feedback.

Related JIRA tickets:

- [IA-2494](https://bluesquare.atlassian.net/browse/IA-2494)
- [IA-2229](https://bluesquare.atlassian.net/browse/IA-2229) (Epic)

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- [x] My migrations file are included
- [x] Are there enough tests
- [x] Documentation has been included (for new feature)

## Changes

- Add validation checks:
    - check that `new_org_unit_type` is part of the user account
    - check that `new_parent` and `org_unit` have the same version
    - check that `new_parent` is not already a child of `org_unit`
- Simplify validation of `approved_fields` by calling the `clean()` method of the model
- Simplify some tests methods names
- Do not rollback in case of `IntegrityError`
- Add a new feature flag `ORG_UNIT_CHANGE_REQUEST`
- Fix filters labels in the UI

[IA-2494]: https://bluesquare.atlassian.net/browse/IA-2494?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[IA-2229]: https://bluesquare.atlassian.net/browse/IA-2229?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: iaso/api/org_unit_change_requests/filters.py [hunk 1]
```diff
@@ -16,11 +17,11 @@ class Meta:
...
-    org_unit_id = django_filters.NumberFilter(field_name="org_unit_id")
-    org_unit_type_id = django_filters.NumberFilter(method="filter_org_unit_type_id")
-    parent_id = django_filters.NumberFilter(method="filter_parent_id")
-    groups = django_filters.CharFilter(method="filter_groups")
-    project = django_filters.NumberFilter(field_name="org_unit__org_unit_type__projects")
+    org_unit_id = django_filters.NumberFilter(field_name="org_unit_id", label=_("Org unit ID"))
+    org_unit_type_id = django_filters.NumberFilter(method="filter_org_unit_type_id", label=_("Org unit type ID"))
+    parent_id = django_filters.NumberFilter(method="filter_parent_id", label=_("Parent ID"))
+    groups = django_filters.CharFilter(method="filter_groups", label=_("Groups IDs (comma-separated)"))
+    project = django_filters.NumberFilter(field_name="org_unit__org_unit_type__projects", label=_("Project ID"))
```
Comment (kemar) [inline iaso/api/org_unit_change_requests/filters.py:24]: This fixes the UI:

filters

File: iaso/api/org_unit_change_requests/serializers.py [hunk 2]
```diff
@@ -248,22 +260,18 @@ def validate_status(self, value):
...
 
-    def validate_approved_fields(self, value):
-        try:
-            OrgUnitChangeRequest.clean_approved_fields(value)
-        except ValidationError as e:
-            raise serializers.ValidationError(e.messages[0]) from e
-        return value
+    def validate(self, validated_data):
+        instance = OrgUnitChangeRequest(**validated_data)
+        instance.clean()
```
Comment (kemar) [inline iaso/api/org_unit_change_requests/serializers.py:265]: Calling `clean()` from the model here allows to delete `def validate_approved_fields()` and to get rid of redundant code.

File: iaso/migrations/0239_add_feature_flag_ORG_UNIT_CHANGE_REQUEST.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    FeatureFlag = apps.get_model("iaso", "FeatureFlag")
+    FeatureFlag.objects.filter(code="ORG_UNIT_CHANGE_REQUEST").delete()
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("iaso", "0238_rename_new_accuracy_orgunitchangerequest_new_location_accuracy_uuid"),
+    ]
+
+    operations = [migrations.RunPython(create_feature_flags, destroy_feature_flags)]
```
Comment (kemar) [inline iaso/migrations/0239_add_feature_flag_ORG_UNIT_CHANGE_REQUEST.py:24]: Should the feature flag be used by the backend code? Or is it only for clients usage?

File: iaso/models/org_unit.py [hunk 4]
```diff
@@ -705,9 +704,3 @@ def get_new_fields(cls) -> typing.List[str]:
         Returns the list of fields names which can store a change request.
         """
         return [field.name for field in cls._meta.get_fields() if field.name.startswith("new_")]
-
-    @classmethod
-    def clean_approved_fields(cls, approved_fields):
-        for name in approved_fields:
-            if name not in cls.get_new_fields():
-                raise ValidationError({"approved_fields": f"Value {name} is not a valid choice."})
```
Comment (kemar) [inline iaso/models/org_unit.py:713]: Calling `clean()` from the serializer validation mechanism allows to simplify things here.

File: iaso/api/org_unit_change_requests/serializers.py [hunk 5]
```diff
@@ -203,29 +202,42 @@ class Meta:
...
-        ]
-        if not any([attrs.get(field) for field in new_fields]):
+    def validate_new_org_unit_type_id(self, new_org_unit_type):
+        request = self.context.get("request")
+        if request and not new_org_unit_type.projects.filter(account=request.user.iaso_profile.account).exists():
+            raise serializers.ValidationError("`new_org_unit_type_id` is not part of the user account.")
+        return new_org_unit_type
+
+    def validate(self, validated_data):
+        # Fields names are different between API and model, e.g. `new_parent_id` VS `new_parent`.
```
Comment (madewulf) [inline iaso/api/org_unit_change_requests/serializers.py:221]: Wondering if that comment really applies here. Isn't it stale?

File: iaso/migrations/0239_add_feature_flag_ORG_UNIT_CHANGE_REQUEST.py (hunk 3 above)
Comment (madewulf) [reply iaso/migrations/0239_add_feature_flag_ORG_UNIT_CHANGE_REQUEST.py:24]: It's good to be able to have the feature activated only on some projects (= some apps), so it's ok like it is now. If we want to hide the feature on web too, we'll desactivate the module for the account (the current PR of Franck)

**Review by madewulf — APPROVED** [review]

**Pull request merged**